File: node-lib/src/protocol.ts
import { Buffer } from "node:buffer";

// ---------------------------------------------------------------------------
// Stream meta constants
// ---------------------------------------------------------------------------

export const SESSION_PACKET_FLAG = 1n << 63n;
export const CONFIG_PACKET_FLAG = 1n << 62n;
export const KEY_FRAME_FLAG = 1n << 61n;
export const PTS_MASK = (1n << 61n) - 1n;

export enum VideoCodec {
  H264 = "h264",
  H265 = "h265",
  AV1 = "av1",
}

export enum AudioCodec {
  OPUS = "opus",
  AAC = "aac",
  FLAC = "flac",
  RAW = "raw",
}

function getCodecId(name: string): number {
  const buf = Buffer.alloc(4, 0);
  const nameBuf = Buffer.from(name, "ascii");
  nameBuf.copy(buf, 4 - nameBuf.length);
  return buf.readUInt32BE(0);
}

export const VIDEO_CODEC_IDS: Record<number, VideoCodec> = Object.fromEntries(
  Object.values(VideoCodec).map((c) => [getCodecId(c), c])
);

export const AUDIO_CODEC_IDS: Record<number, AudioCodec> = Object.fromEntries(
  Object.values(AudioCodec).map((c) => [getCodecId(c), c])
);

// ---------------------------------------------------------------------------
// Frame headers
// ---------------------------------------------------------------------------

export interface SessionPacket {
  kind: "session";
  width: number;
  height: number;
  clientResized: boolean;
}

export interface FrameMeta {
  kind: "frame";
  ptsUs: bigint;
  size: number;
  config: boolean;
  keyFrame: boolean;
}

export function parseFrameHeader(header: Buffer): SessionPacket | FrameMeta {
  if (header.length !== 12) {
    throw new Error(`frame header must be 12 bytes, got ${header.length}`);
  }

  const ptsAndFlags = header.readBigUInt64BE(0);
  const size = header.readUInt32BE(8);

  if (ptsAndFlags & SESSION_PACKET_FLAG) {
    const flagsHigh = Number((ptsAndFlags >> 32n) & 0xffffffffn);
    const width = Number(ptsAndFlags & 0xffffffffn);
    const height = size;
    return {
      kind: "session",
      width,
      height,
      clientResized: (flagsHigh & 0x01) !== 0,
    };
  }

  return {
    kind: "frame",
    ptsUs: ptsAndFlags & PTS_MASK,
    size,
    config: (ptsAndFlags & CONFIG_PACKET_FLAG) !== 0n,
    keyFrame: (ptsAndFlags & KEY_FRAME_FLAG) !== 0n,
  };
}

// ---------------------------------------------------------------------------
// Control messages (client -> device)
// ---------------------------------------------------------------------------

export enum ControlMessageType {
  INJECT_KEYCODE = 0,
  INJECT_TEXT = 1,
  INJECT_TOUCH_EVENT = 2,
  INJECT_SCROLL_EVENT = 3,
  BACK_OR_SCREEN_ON = 4,
  EXPAND_NOTIFICATION_PANEL = 5,
  EXPAND_SETTINGS_PANEL = 6,
  COLLAPSE_PANELS = 7,
  GET_CLIPBOARD = 8,
  SET_CLIPBOARD = 9,
  SET_DISPLAY_POWER = 10,
  ROTATE_DEVICE = 11,
  UHID_CREATE = 12,
  UHID_INPUT = 13,
  UHID_DESTROY = 14,
  OPEN_HARD_KEYBOARD_SETTINGS = 15,
  START_APP = 16,
  RESET_VIDEO = 17,
  CAMERA_SET_TORCH = 18,
  CAMERA_ZOOM_IN = 19,
  CAMERA_ZOOM_OUT = 20,
  RESIZE_DISPLAY = 21,
}

export const ACTION_DOWN = 0;
export const ACTION_UP = 1;
export const ACTION_MOVE = 2;

export const KEY_ACTION_DOWN = 0;
export const KEY_ACTION_UP = 1;

export const BUTTON_PRIMARY = 1 << 0;
export const BUTTON_SECONDARY = 1 << 1;
export const BUTTON_TERTIARY = 1 << 2;

export const POINTER_ID_MOUSE = -1n; // 0xFFFF_FFFF_FFFF_FFFF as signed i64
export const POINTER_ID_GENERIC_FINGER = -2n; // 0xFFFF_FFFF_FFFF_FFFE

export const INJECT_TEXT_MAX_LENGTH = 300;
export const SET_CLIPBOARD_TEXT_MAX_LENGTH = (1 << 18) - 1;

function u16FixedPoint(value: number): number {
  if (value >= 1.0) return 0xffff;
  if (value <= 0.0) return 0;
  return Math.floor(value * 0x10000) & 0xffff;
}

function i16FixedPoint(value: number): number {
  if (value >= 1.0) return 0x7fff;
  if (value <= -1.0) return 0x8000; // stored as unsigned 16-bit
  return Math.floor(value * 0x8000) & 0xffff;
}

export interface ControlMessage {
  type: ControlMessageType;
  action?: number;
  keycode?: number;
  repeat?: number;
  metaState?: number;
  pointerId?: bigint;
  pressure?: number;
  actionButton?: number;
  buttons?: number;
  x?: number;
  y?: number;
  screenWidth?: number;
  screenHeight?: number;
  hScroll?: number;
  vScroll?: number;
  text?: string;
  paste?: boolean;
  sequence?: bigint;
  copyKey?: number;
  on?: boolean;
  width?: number;
  height?: number;
}

export function serializeControlMessage(msg: ControlMessage): Buffer {
  const type = msg.type;
  switch (type) {
    case ControlMessageType.INJECT_KEYCODE: {
      const buf = Buffer.alloc(14);
      buf.writeUInt8(type, 0);
      buf.writeUInt8((msg.action || 0) & 0xff, 1);
      buf.writeUInt32BE(msg.keycode || 0, 2);
      buf.writeUInt32BE(msg.repeat || 0, 6);
      buf.writeUInt32BE(msg.metaState || 0, 10);
      return buf;
    }
    case ControlMessageType.INJECT_TEXT: {
      const textBuf = truncateUtf8(msg.text || "", INJECT_TEXT_MAX_LENGTH);
      const buf = Buffer.alloc(5 + textBuf.length);
      buf.writeUInt8(type, 0);
      buf.writeUInt32BE(textBuf.length, 1);
      textBuf.copy(buf, 5);
      return buf;
    }
    case ControlMessageType.INJECT_TOUCH_EVENT: {
      const buf = Buffer.alloc(32);
      buf.writeUInt8(type, 0);
      buf.writeUInt8((msg.action || 0) & 0xff, 1);
      buf.writeBigUInt64BE(BigInt.asUintN(64, msg.pointerId || 0n), 2);
      buf.writeInt32BE(Math.floor(msg.x || 0), 10);
      buf.writeInt32BE(Math.floor(msg.y || 0), 14);
      buf.writeUInt16BE((msg.screenWidth || 0) & 0xffff, 18);
      buf.writeUInt16BE((msg.screenHeight || 0) & 0xffff, 20);
      buf.writeUInt16BE(u16FixedPoint(msg.pressure ?? 1.0), 22);
      buf.writeUInt32BE((msg.actionButton || 0) & 0xffffffff, 24);
      buf.writeUInt32BE((msg.buttons || 0) & 0xffffffff, 28);
      return buf;
    }
    case ControlMessageType.INJECT_SCROLL_EVENT: {
      const buf = Buffer.alloc(21);
      buf.writeUInt8(type, 0);
      buf.writeInt32BE(Math.floor(msg.x || 0), 1);
      buf.writeInt32BE(Math.floor(msg.y || 0), 5);
      buf.writeUInt16BE((msg.screenWidth || 0) & 0xffff, 9);
      buf.writeUInt16BE((msg.screenHeight || 0) & 0xffff, 11);
      const h = Math.max(-1.0, Math.min(1.0, (msg.hScroll || 0.0) / 16.0));
      const v = Math.max(-1.0, Math.min(1.0, (msg.vScroll || 0.0) / 16.0));
      buf.writeUInt16BE(i16FixedPoint(h), 13);
      buf.writeUInt16BE(i16FixedPoint(v), 15);
      buf.writeUInt32BE((msg.buttons || 0) & 0xffffffff, 17);
      return buf;
    }
    case ControlMessageType.BACK_OR_SCREEN_ON: {
      const buf = Buffer.alloc(2);
      buf.writeUInt8(type, 0);
      buf.writeUInt8((msg.action || 0) & 0xff, 1);
      return buf;
    }
    case ControlMessageType.SET_CLIPBOARD: {
      const textBuf = truncateUtf8(
        msg.text || "",
        SET_CLIPBOARD_TEXT_MAX_LENGTH
      );
      const buf = Buffer.alloc(14 + textBuf.length);
      buf.writeUInt8(type, 0);
      buf.writeBigUInt64BE(BigInt.asUintN(64, msg.sequence || 0n), 1);
      buf.writeUInt8(msg.paste ? 1 : 0, 9);
      buf.writeUInt32BE(textBuf.length, 10);
      textBuf.copy(buf, 14);
      return buf;
    }
    case ControlMessageType.SET_DISPLAY_POWER:
    case ControlMessageType.CAMERA_SET_TORCH: {
      const buf = Buffer.alloc(2);
      buf.writeUInt8(type, 0);
      buf.writeUInt8(msg.on ? 1 : 0, 1);
      return buf;
    }
    case ControlMessageType.RESIZE_DISPLAY: {
      const buf = Buffer.alloc(5);
      buf.writeUInt8(type, 0);
      buf.writeUInt16BE((msg.width || 0) & 0xffff, 1);
      buf.writeUInt16BE((msg.height || 0) & 0xffff, 3);
      return buf;
    }
    case ControlMessageType.EXPAND_NOTIFICATION_PANEL:
    case ControlMessageType.EXPAND_SETTINGS_PANEL:
    case ControlMessageType.COLLAPSE_PANELS:
    case ControlMessageType.ROTATE_DEVICE:
    case ControlMessageType.OPEN_HARD_KEYBOARD_SETTINGS:
    case ControlMessageType.RESET_VIDEO:
    case ControlMessageType.CAMERA_ZOOM_IN:
    case ControlMessageType.CAMERA_ZOOM_OUT: {
      const buf = Buffer.alloc(1);
      buf.writeUInt8(type, 0);
      return buf;
    }
    default:
      throw new Error(`unsupported control message type: ${type}`);
  }
}

function truncateUtf8(text: string, maxBytes: number): Buffer {
  const buf = Buffer.from(text, "utf-8");
  if (buf.length <= maxBytes) return buf;

  let end = maxBytes;
  // If the last byte is a continuation byte (10xxxxxx), move back
  while (end > 0 && (buf[end] & 0xc0) === 0x80) {
    end--;
  }
  // If the last byte is a leading byte of a multi-byte sequence, drop it
  if (end > 0 && (buf[end] & 0x80) !== 0) {
    end--;
  }
  return buf.subarray(0, end);
}

// ---------------------------------------------------------------------------
// Device messages (device -> client)
// ---------------------------------------------------------------------------

export enum DeviceMessageType {
  CLIPBOARD = 0,
  ACK_CLIPBOARD = 1,
  UHID_OUTPUT = 2,
}

export interface DeviceMessage {
  type: DeviceMessageType;
  text?: string;
  sequence?: bigint;
  uhidId?: number;
  data?: Buffer;
}

export async function parseDeviceMessage(
  readExact: (n: number) => Promise<Buffer>
): Promise<DeviceMessage> {
  const head = await readExact(1);
  const type = head[0] as DeviceMessageType;

  switch (type) {
    case DeviceMessageType.CLIPBOARD: {
      const lenBuf = await readExact(4);
      const length = lenBuf.readUInt32BE(0);
      const textBuf = await readExact(length);
      return { type, text: textBuf.toString("utf-8") };
    }
    case DeviceMessageType.ACK_CLIPBOARD: {
      const seqBuf = await readExact(8);
      return { type, sequence: seqBuf.readBigUInt64BE(0) };
    }
    case DeviceMessageType.UHID_OUTPUT: {
      const metaBuf = await readExact(4);
      const uhidId = metaBuf.readUInt16BE(0);
      const length = metaBuf.readUInt16BE(2);
      const data = await readExact(length);
      return { type, uhidId, data };
    }
    default:
      throw new Error(`unknown device message type: ${type}`);
  }
}

File: src/android_stream/scrcpy_v4/protocol.py
"""Pure-Python helpers for the scrcpy v4.0 wire protocol.

Only what we need for a browser bridge:

- parse the ``codec id`` (4 bytes) preamble on the video / audio sockets
- parse the 12-byte session packet (video only) and the 12-byte frame header
- serialize :class:`ControlMessage` instances back into bytes the scrcpy
  server understands

The byte layouts mirror exactly what the official Java server reads/writes.
The ground truth lives in ``scrcpy/server/src/main/java/com/genymobile/scrcpy/``
- :class:`Streamer.writeFrameMeta` (video / audio packet header)
- :class:`Streamer.writeSessionMeta` (video-only session header)
- :class:`ControlMessageReader` (control message deserializer)
"""

from __future__ import annotations

import enum
import struct
from dataclasses import dataclass
from typing import Final


# ---------------------------------------------------------------------------
# Stream meta constants
# ---------------------------------------------------------------------------

# Top three bits of the 8-byte ``ptsAndFlags`` header.
SESSION_PACKET_FLAG: Final[int] = 1 << 63
CONFIG_PACKET_FLAG: Final[int] = 1 << 62
KEY_FRAME_FLAG: Final[int] = 1 << 61
PTS_MASK: Final[int] = (1 << 61) - 1


class VideoCodec(str, enum.Enum):
    H264 = "h264"
    H265 = "h265"
    AV1 = "av1"


class AudioCodec(str, enum.Enum):
    OPUS = "opus"
    AAC = "aac"
    FLAC = "flac"
    RAW = "raw"


def _codec_id(name: str) -> int:
    raw = name.encode("ascii")
    if len(raw) > 4:
        raise ValueError(f"codec name too long: {name!r}")
    raw = raw.rjust(4, b"\x00")
    return int.from_bytes(raw, "big")


VIDEO_CODEC_IDS: Final[dict[int, VideoCodec]] = {
    _codec_id(c.value): c for c in VideoCodec
}
AUDIO_CODEC_IDS: Final[dict[int, AudioCodec]] = {
    _codec_id(c.value): c for c in AudioCodec
}


# ---------------------------------------------------------------------------
# Frame headers
# ---------------------------------------------------------------------------


@dataclass(slots=True, frozen=True)
class SessionPacket:
    width: int
    height: int
    client_resized: bool


@dataclass(slots=True, frozen=True)
class FrameMeta:
    pts_us: int
    size: int
    config: bool
    key_frame: bool


def parse_frame_header(header: bytes) -> SessionPacket | FrameMeta:
    """Parse one 12-byte stream header from the video/audio socket.

    The video stream may interleave *session* packets (signalled by the MSB
    of the first byte) and *media* packets. Audio streams contain only media
    packets.
    """
    if len(header) != 12:
        raise ValueError(f"frame header must be 12 bytes, got {len(header)}")

    pts_and_flags, size = struct.unpack(">QI", header)
    if pts_and_flags & SESSION_PACKET_FLAG:
        # Session packet: 4-byte flags + 4-byte width + 4-byte height. The
        # ``size`` field of the standard header doubles as the video height.
        flags_high = (pts_and_flags >> 32) & 0xFFFFFFFF
        width = pts_and_flags & 0xFFFFFFFF
        height = size
        return SessionPacket(
            width=width,
            height=height,
            client_resized=bool(flags_high & 0x1),
        )

    return FrameMeta(
        pts_us=pts_and_flags & PTS_MASK,
        size=size,
        config=bool(pts_and_flags & CONFIG_PACKET_FLAG),
        key_frame=bool(pts_and_flags & KEY_FRAME_FLAG),
    )


# ---------------------------------------------------------------------------
# Control messages (client -> device)
# ---------------------------------------------------------------------------


class CONTROL_MSG_TYPE(enum.IntEnum):
    INJECT_KEYCODE = 0
    INJECT_TEXT = 1
    INJECT_TOUCH_EVENT = 2
    INJECT_SCROLL_EVENT = 3
    BACK_OR_SCREEN_ON = 4
    EXPAND_NOTIFICATION_PANEL = 5
    EXPAND_SETTINGS_PANEL = 6
    COLLAPSE_PANELS = 7
    GET_CLIPBOARD = 8
    SET_CLIPBOARD = 9
    SET_DISPLAY_POWER = 10
    ROTATE_DEVICE = 11
    UHID_CREATE = 12
    UHID_INPUT = 13
    UHID_DESTROY = 14
    OPEN_HARD_KEYBOARD_SETTINGS = 15
    START_APP = 16
    RESET_VIDEO = 17
    CAMERA_SET_TORCH = 18
    CAMERA_ZOOM_IN = 19
    CAMERA_ZOOM_OUT = 20
    RESIZE_DISPLAY = 21


# Same constants as android.view.MotionEvent.ACTION_*.
ACTION_DOWN: Final[int] = 0
ACTION_UP: Final[int] = 1
ACTION_MOVE: Final[int] = 2

# android.view.KeyEvent.ACTION_*.
KEY_ACTION_DOWN: Final[int] = 0
KEY_ACTION_UP: Final[int] = 1

# android.view.MotionEvent.BUTTON_*.
BUTTON_PRIMARY: Final[int] = 1 << 0
BUTTON_SECONDARY: Final[int] = 1 << 1
BUTTON_TERTIARY: Final[int] = 1 << 2

# Well-known pointer ids used by scrcpy. Anything else can be used for
# secondary fingers.
POINTER_ID_MOUSE: Final[int] = 0xFFFF_FFFF_FFFF_FFFF
POINTER_ID_GENERIC_FINGER: Final[int] = 0xFFFF_FFFF_FFFF_FFFE

INJECT_TEXT_MAX_LENGTH: Final[int] = 300
SET_CLIPBOARD_TEXT_MAX_LENGTH: Final[int] = (1 << 18) - 14


def _u16_fixed_point(value: float) -> int:
    """Convert ``[0.0, 1.0]`` to the wire 16-bit fixed-point format."""
    if value >= 1.0:
        return 0xFFFF
    if value <= 0.0:
        return 0
    return int(value * 0x1_0000) & 0xFFFF


def _i16_fixed_point(value: float) -> int:
    """Convert ``[-1.0, 1.0]`` to the wire 16-bit fixed-point format."""
    if value >= 1.0:
        return 0x7FFF
    if value <= -1.0:
        return -0x8000 & 0xFFFF
    return int(value * 0x8000) & 0xFFFF


@dataclass(slots=True)
class ControlMessage:
    """Lightweight equivalent of the Java ``ControlMessage``.

    Construct via the ``create_*`` class methods, then call
    :func:`serialize_control_message` to obtain the wire bytes.
    """

    type: CONTROL_MSG_TYPE
    # All other fields are optional and message-type specific.
    action: int = 0
    keycode: int = 0
    repeat: int = 0
    meta_state: int = 0
    pointer_id: int = 0
    pressure: float = 1.0
    action_button: int = 0
    buttons: int = 0
    x: int = 0
    y: int = 0
    screen_width: int = 0
    screen_height: int = 0
    h_scroll: float = 0.0
    v_scroll: float = 0.0
    text: str = ""
    paste: bool = False
    sequence: int = 0
    copy_key: int = 0
    on: bool = False
    width: int = 0
    height: int = 0

    # ----- factories ------------------------------------------------------
    @classmethod
    def inject_keycode(
        cls,
        action: int,
        keycode: int,
        repeat: int = 0,
        meta_state: int = 0,
    ) -> "ControlMessage":
        return cls(
            type=CONTROL_MSG_TYPE.INJECT_KEYCODE,
            action=action,
            keycode=keycode,
            repeat=repeat,
            meta_state=meta_state,
        )

    @classmethod
    def inject_text(cls, text: str) -> "ControlMessage":
        return cls(type=CONTROL_MSG_TYPE.INJECT_TEXT, text=text)

    @classmethod
    def inject_touch_event(
        cls,
        *,
        action: int,
        pointer_id: int,
        x: int,
        y: int,
        screen_width: int,
        screen_height: int,
        pressure: float = 1.0,
        action_button: int = 0,
        buttons: int = 0,
    ) -> "ControlMessage":
        return cls(
            type=CONTROL_MSG_TYPE.INJECT_TOUCH_EVENT,
            action=action,
            pointer_id=pointer_id,
            x=x,
            y=y,
            screen_width=screen_width,
            screen_height=screen_height,
            pressure=pressure,
            action_button=action_button,
            buttons=buttons,
        )

    @classmethod
    def inject_scroll_event(
        cls,
        *,
        x: int,
        y: int,
        screen_width: int,
        screen_height: int,
        h_scroll: float,
        v_scroll: float,
        buttons: int = 0,
    ) -> "ControlMessage":
        return cls(
            type=CONTROL_MSG_TYPE.INJECT_SCROLL_EVENT,
            x=x,
            y=y,
            screen_width=screen_width,
            screen_height=screen_height,
            h_scroll=h_scroll,
            v_scroll=v_scroll,
            buttons=buttons,
        )

    @classmethod
    def back_or_screen_on(cls, action: int) -> "ControlMessage":
        return cls(type=CONTROL_MSG_TYPE.BACK_OR_SCREEN_ON, action=action)

    @classmethod
    def empty(cls, msg_type: CONTROL_MSG_TYPE) -> "ControlMessage":
        return cls(type=msg_type)

    @classmethod
    def set_clipboard(
        cls, *, sequence: int, text: str, paste: bool
    ) -> "ControlMessage":
        return cls(
            type=CONTROL_MSG_TYPE.SET_CLIPBOARD,
            sequence=sequence,
            text=text,
            paste=paste,
        )

    @classmethod
    def set_display_power(cls, on: bool) -> "ControlMessage":
        return cls(type=CONTROL_MSG_TYPE.SET_DISPLAY_POWER, on=on)


def _truncate_utf8(text: str, max_bytes: int) -> bytes:
    """Truncate ``text`` so its UTF-8 encoding fits in ``max_bytes`` bytes.

    The Java server (``StringUtils.getUtf8TruncationIndex``) enforces this
    same limit and silently truncates beyond it. We match that behavior so
    the bytes we put on the wire are always valid UTF-8 even when the input
    is too long. The trailing partial character (if any) is dropped — never
    half a multi-byte sequence.
    """
    raw = text.encode("utf-8")
    if len(raw) <= max_bytes:
        return raw
    truncated = bytearray(raw[:max_bytes])
    # Walk back over any continuation bytes (top bits ``10``) and the leading
    # byte that introduced the partial sequence (top bits ``11``). Pure ASCII
    # boundaries (top bit clear) are always safe.
    while truncated:
        last = truncated[-1]
        if last < 0x80:
            break
        if (last & 0xC0) == 0x80:
            truncated.pop()
            continue
        # Leading byte of an incomplete multi-byte sequence — drop it too.
        truncated.pop()
        break
    return bytes(truncated)


def serialize_control_message(msg: ControlMessage) -> bytes:
    """Serialize ``msg`` into the binary form expected by the device server."""
    t = msg.type
    if t == CONTROL_MSG_TYPE.INJECT_KEYCODE:
        return struct.pack(
            ">BBIII", t, msg.action & 0xFF, msg.keycode, msg.repeat, msg.meta_state
        )
    if t == CONTROL_MSG_TYPE.INJECT_TEXT:
        raw = _truncate_utf8(msg.text, INJECT_TEXT_MAX_LENGTH)
        return struct.pack(">BI", t, len(raw)) + raw
    if t == CONTROL_MSG_TYPE.INJECT_TOUCH_EVENT:
        # Position payload: x(i32 BE) y(i32 BE) sw(u16 BE) sh(u16 BE).
        position = struct.pack(
            ">iiHH",
            int(msg.x),
            int(msg.y),
            msg.screen_width & 0xFFFF,
            msg.screen_height & 0xFFFF,
        )
        return (
            struct.pack(">BBQ", t, msg.action & 0xFF, msg.pointer_id & 0xFFFFFFFFFFFFFFFF)
            + position
            + struct.pack(
                ">HII",
                _u16_fixed_point(msg.pressure),
                msg.action_button & 0xFFFFFFFF,
                msg.buttons & 0xFFFFFFFF,
            )
        )
    if t == CONTROL_MSG_TYPE.INJECT_SCROLL_EVENT:
        position = struct.pack(
            ">iiHH",
            int(msg.x),
            int(msg.y),
            msg.screen_width & 0xFFFF,
            msg.screen_height & 0xFFFF,
        )
        # Wire range is [-1, 1], mapped from a [-16, 16] application-level
        # range. We mirror the C client's CLAMP and division by 16.
        h = max(-1.0, min(1.0, msg.h_scroll / 16.0))
        v = max(-1.0, min(1.0, msg.v_scroll / 16.0))
        return (
            struct.pack(">B", t)
            + position
            + struct.pack(
                ">HHI",
                _i16_fixed_point(h),
                _i16_fixed_point(v),
                msg.buttons & 0xFFFFFFFF,
            )
        )
    if t == CONTROL_MSG_TYPE.BACK_OR_SCREEN_ON:
        return struct.pack(">BB", t, msg.action & 0xFF)
    if t == CONTROL_MSG_TYPE.GET_CLIPBOARD:
        return struct.pack(">BB", t, msg.copy_key & 0xFF)
    if t == CONTROL_MSG_TYPE.SET_CLIPBOARD:
        raw = _truncate_utf8(msg.text, SET_CLIPBOARD_TEXT_MAX_LENGTH)
        return (
            struct.pack(">BQB", t, msg.sequence & 0xFFFFFFFFFFFFFFFF, 1 if msg.paste else 0)
            + struct.pack(">I", len(raw))
            + raw
        )
    if t == CONTROL_MSG_TYPE.SET_DISPLAY_POWER:
        return struct.pack(">BB", t, 1 if msg.on else 0)
    if t == CONTROL_MSG_TYPE.CAMERA_SET_TORCH:
        return struct.pack(">BB", t, 1 if msg.on else 0)
    if t == CONTROL_MSG_TYPE.RESIZE_DISPLAY:
        return struct.pack(
            ">BHH", t, msg.width & 0xFFFF, msg.height & 0xFFFF
        )
    if t in (
        CONTROL_MSG_TYPE.EXPAND_NOTIFICATION_PANEL,
        CONTROL_MSG_TYPE.EXPAND_SETTINGS_PANEL,
        CONTROL_MSG_TYPE.COLLAPSE_PANELS,
        CONTROL_MSG_TYPE.ROTATE_DEVICE,
        CONTROL_MSG_TYPE.OPEN_HARD_KEYBOARD_SETTINGS,
        CONTROL_MSG_TYPE.RESET_VIDEO,
        CONTROL_MSG_TYPE.CAMERA_ZOOM_IN,
        CONTROL_MSG_TYPE.CAMERA_ZOOM_OUT,
    ):
        return struct.pack(">B", t)

    raise ValueError(f"unsupported control message type: {t}")


# ---------------------------------------------------------------------------
# Device messages (device -> client)
# ---------------------------------------------------------------------------


class DEVICE_MSG_TYPE(enum.IntEnum):
    CLIPBOARD = 0
    ACK_CLIPBOARD = 1
    UHID_OUTPUT = 2


@dataclass(slots=True, frozen=True)
class DeviceMessage:
    type: DEVICE_MSG_TYPE
    text: str | None = None
    sequence: int | None = None
    uhid_id: int | None = None
    data: bytes | None = None


def parse_device_message(read_exact) -> DeviceMessage:
    """Parse one device message using a blocking ``read_exact(n) -> bytes``.

    The reader callable is expected to either return exactly ``n`` bytes or
    raise an :class:`OSError`-like exception on disconnect.
    """
    head = read_exact(1)
    msg_type = DEVICE_MSG_TYPE(head[0])
    if msg_type == DEVICE_MSG_TYPE.CLIPBOARD:
        (length,) = struct.unpack(">I", read_exact(4))
        text = read_exact(length).decode("utf-8", errors="replace")
        return DeviceMessage(type=msg_type, text=text)
    if msg_type == DEVICE_MSG_TYPE.ACK_CLIPBOARD:
        (sequence,) = struct.unpack(">Q", read_exact(8))
        return DeviceMessage(type=msg_type, sequence=sequence)
    if msg_type == DEVICE_MSG_TYPE.UHID_OUTPUT:
        uhid_id, length = struct.unpack(">HH", read_exact(4))
        return DeviceMessage(
            type=msg_type, uhid_id=uhid_id, data=read_exact(length)
        )
    raise ValueError(f"unknown device message type: {msg_type}")
